codec-haproxy/src/main/java/io/netty/contrib/handler/codec/haproxy/HAProxyMessageEncoder.java

```diff
@@ -86,47 +91,57 @@ private static void encodeV2(HAProxyMessage msg, ByteBuf out) {
+                out.writeShort((short) msg.destinationPort());
                 encodeTlvs(msg.tlvs(), out);
                 break;
             case AF_UNIX:
-                out.writeShort(TOTAL_UNIX_ADDRESS_BYTES_LENGTH + msg.tlvNumBytes());
-                int srcAddrBytesWritten = out.writeCharSequence(msg.sourceAddress(), CharsetUtil.US_ASCII);
-                out.writeZero(UNIX_ADDRESS_BYTES_LENGTH - srcAddrBytesWritten);
-                int dstAddrBytesWritten = out.writeCharSequence(msg.destinationAddress(), CharsetUtil.US_ASCII);
-                out.writeZero(UNIX_ADDRESS_BYTES_LENGTH - dstAddrBytesWritten);
+                out.writeShort((short) (TOTAL_UNIX_ADDRESS_BYTES_LENGTH + msg.tlvNumBytes()));
+                int srcAddrBytesWritten = out.writerOffset();
+                out.writeCharSequence(msg.sourceAddress(), CharsetUtil.US_ASCII);
+                srcAddrBytesWritten = out.writerOffset() - srcAddrBytesWritten;
+                int length = UNIX_ADDRESS_BYTES_LENGTH - srcAddrBytesWritten;
+                for(int i = 0; i < length; ++i) {
+                    out.writeByte((byte) 0);
+                }
+                int dstAddrBytesWritten = out.writerOffset();
+                out.writeCharSequence(msg.destinationAddress(), CharsetUtil.US_ASCII);
+                dstAddrBytesWritten = out.writerOffset() - dstAddrBytesWritten;
+                length = UNIX_ADDRESS_BYTES_LENGTH - dstAddrBytesWritten;
+                for(int i = 0; i < length; ++i) {
+                    out.writeByte((byte) 0);
+                }
                 encodeTlvs(msg.tlvs(), out);
                 break;
             case AF_UNSPEC:
-                out.writeShort(0);
+                out.writeShort((short) 0);
                 break;
             default:
                 throw new HAProxyProtocolException("unexpected addrFamily");
         }
     }
 
-    private static void encodeTlv(HAProxyTLV haProxyTLV, ByteBuf out) {
+    private static void encodeTlv(HAProxyTLV haProxyTLV, Buffer out) {
         if (haProxyTLV instanceof HAProxySSLTLV) {
             HAProxySSLTLV ssltlv = (HAProxySSLTLV) haProxyTLV;
             out.writeByte(haProxyTLV.typeByteValue());
-            out.writeShort(ssltlv.contentNumBytes());
+            out.writeShort((short) ssltlv.contentNumBytes());
             out.writeByte(ssltlv.client());
             out.writeInt(ssltlv.verify());
             encodeTlvs(ssltlv.encapsulatedTLVs(), out);
         } else {
             out.writeByte(haProxyTLV.typeByteValue());
-            ByteBuf value = haProxyTLV.content();
+            Buffer value = haProxyTLV.content();
             int readableBytes = value.readableBytes();
-            out.writeShort(readableBytes);
-            out.writeBytes(value.readSlice(readableBytes));
+            out.writeShort((short) readableBytes);
+            out.writeBytes(value.copy(value.readerOffset(), readableBytes));
```

Comment:
The `Buffer` instance from `copy()` isn't closed.
We can instead write something like

```java
value.copyInto(value.readerOffset(), out, out.readerOffset(), readableBytes);
out.skipWritable(readableBytes);
```

Reply:
fixed